axelor-production/src/main/java/com/axelor/apps/production/service/productionorder/ProductionOrderSaleOrderServiceImpl.java

```diff
@@ -59,38 +66,127 @@ public ProductionOrderSaleOrderServiceImpl(
 
   @Override
   @Transactional(rollbackOn = {AxelorException.class})
-  public List<Long> generateProductionOrder(SaleOrder saleOrder) throws AxelorException {
+  public Set<Long> generateProductionOrder(SaleOrder saleOrder) throws AxelorException {
 
     boolean oneProdOrderPerSO = appProductionService.getAppProduction().getOneProdOrderPerSO();
 
-    List<Long> productionOrderIdList = new ArrayList<>();
+    Set<Long> productionOrderIds = new HashSet<>();
     if (saleOrder.getSaleOrderLineList() == null) {
-      return productionOrderIdList;
+      return productionOrderIds;
     }
 
-    ProductionOrder productionOrder = null;
-    for (SaleOrderLine saleOrderLine : saleOrder.getSaleOrderLineList()) {
+    List<SaleOrderLine> saleOrderLineList =
+        saleOrder.getSaleOrderLineList().stream()
+            .filter(line -> CollectionUtils.isEmpty(line.getManufOrderList()))
+            .collect(Collectors.toList());
+
+    if (oneProdOrderPerSO) {
+      ProductionOrder productionOrder = generateOnePoPerSaleOrder(saleOrder, saleOrderLineList);
+      productionOrderIds.add(productionOrder.getId());
+    } else {
+      List<ProductionOrder> productionOrderList = generateOnePoPerSol(saleOrder, saleOrderLineList);
+      productionOrderIds.addAll(
+          productionOrderList.stream().map(ProductionOrder::getId).collect(Collectors.toList()));
+      productionOrderIds.addAll(
+          productionOrderRepo
+              .all()
+              .filter("self.saleOrder = :saleOrder")
+              .bind("saleOrder", saleOrder)
+              .fetch()
+              .stream()
+              .map(ProductionOrder::getId)
+              .collect(Collectors.toList()));
+    }
```

Comment:
It looks like a lot of methods are starting with a check on `oneProdOrderPerSO`: maybe it would be easier to have different methods for when `oneProdOrderPerSO` is true or not, and check only once ?